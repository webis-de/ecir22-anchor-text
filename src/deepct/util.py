def clean_qtokens(qtokens):
    ret = {}
    blocklist = load_entrypage_queries()

    for doc, doc_tokens in qtokens.items():
        if len(doc_tokens.keys()) == 0 or len([i for i in doc_tokens.keys() if i in blocklist])/len(doc_tokens.keys()) > 0.2:
            continue
        ret[doc] = doc_tokens

    print('From qtokens with ' + str(len(qtokens)) + ' documents, I filtered terms that may cause train/test overlap so that I now have ' + str(len(ret)) + ' documents.')

    return ret


def load_entrypage_queries():
    from glob import glob
    import importlib
    text_clean = importlib.import_module('train-deep-ct').text_clean

    ret = []
    for f in glob('../../Data/navigational-topics-and-qrels-ms-marco-v1/topics*.tsv'):
        with open(f, 'r') as f:

            for l in f:
                l = l.split('\t')
                if len(l) != 2:
                    raise ValueError('Can not handle ' + str(l))

                ret += text_clean(l[1].strip(), True, True).split(' ')

    return set(ret)


def sample_anchors(anchors_per_doc):
    import json
    import importlib
    text_clean = importlib.import_module('train-deep-ct').text_clean
    blocklist = load_entrypage_queries()
    ret = set()
    for anchors in anchors_per_doc:
        for anchor in set(json.loads(anchors)['anchorText']):
            anchor_terms = text_clean(anchor.strip(), True, True).split(' ')
            if len(anchor_terms) > 0 and len([i for i in anchor_terms if i in blocklist])/len(anchor_terms) < 0.51:
                ret |= set([anchor])

    return ret
