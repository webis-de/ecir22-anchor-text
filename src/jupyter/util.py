import spacy
from copy import deepcopy
import random
import json
import numpy as np
from nltk.stem.porter import PorterStemmer
stemmer = PorterStemmer()
import itertools
nlp = None
ALL_PERMUTATIONS=set([i for i in itertools.permutations([0,0,0,0,0,1,1,1,1,1])])


def distinct_terms(processed_terms):
    return set([i['stemmed'] for i in processed_terms])


def extract_terms(sentence):
    global nlp
    if nlp is None:
        nlp = spacy.load("en_core_web_sm", exclude=["parser", "tagger", "ner", "attribute_ruler", "lemmatizer", "tok2vec"])
    ret = []
    for i in nlp(sentence):
        if not i.is_punct:
            ret += [{'term': str(i), 'stemmed': stemmer.stem(str(i))}]
    return ret


def json_lines(filename):
    from tqdm import tqdm
    import json
    with open(filename, 'r') as f:
        for i in tqdm(f):
            yield json.loads(i)


def enrich_presentation(presentation):
    ret = deepcopy(presentation)
    if 'contents' in ret:
        ret['contentsProcessed'] = extract_terms(ret['contents'])
    if 'anchorText' in ret:
        ret['anchorTextProcessed'] = []
        for anchor in ret['anchorText']:
            ret['anchorTextProcessed'] += [extract_terms(anchor)]

    return ret


def __terms_for_counting(line):
    if 'contentsProcessed' in line:
        yield line['contentsProcessed']
    if 'anchorTextProcessed' in line:
        for anchor in line['anchorTextProcessed']:
            yield anchor


def __flatten(terms):
    ret = {}
    for term in terms.keys():
        term_with_count = [(i,c) for i,c in terms[term].items()]

        most_term = sorted(term_with_count, key=lambda i: i[1], reverse=True)[0][0]
        ret[most_term] = sum(i[1] for i in term_with_count)

    return ret


def count_terms(lines):
    ret = {}
    for line in lines:
        if type(line) == str:
            line = json.loads(line)
        for terms in __terms_for_counting(line):
            for term in terms:
                if term['stemmed'] not in ret:
                    ret[term['stemmed']] = {}
                if term['term'] not in ret[term['stemmed']]:
                    ret[term['stemmed']][term['term']] = 0

                ret[term['stemmed']][term['term']] += 1

    return __flatten(ret)
            
def distribution_distinct_terms(lines):
    ret = {}

    for line in lines:
        if type(line) == str:
            line = json.loads(line)

        terms_in_line = str(len(distinct_terms(line['contentsProcessed'])))
        
        if int(terms_in_line) <= 0:
            continue

        if terms_in_line not in ret:
            ret[terms_in_line] = 0

        to_add = 1
        if 'target_document' in line and type(line['target_document']) == list:
            to_add = len(line['target_document'])
        ret[terms_in_line] += to_add

    return ret

def anchor_text_distribution_distinct_terms(lines, limit=None):
    ret = []
    for line in lines:
        if type(line) == str:
            line = json.loads(line)

        if limit != None and limit< len(line['anchorTextProcessed']):
            continue

        for anchor in line['anchorTextProcessed']:
            ret += [{'contentsProcessed': anchor}]

    return distribution_distinct_terms(ret)

def aggregated_anchor_text_distribution_distinct_terms(lines, limit=None):
    ret = []
    for line in lines:
        if type(line) == str:
            line = json.loads(line)
        contents = []
        if limit != None and limit< len(line['anchorTextProcessed']):
            continue

        for anchor in line['anchorTextProcessed']:
            contents += anchor
        ret += [{'contentsProcessed': contents}]

    return distribution_distinct_terms(ret)

def build_probability_arrays(first, second):
    first_ret = []
    second_ret = []
    first_sum = sum(first.values())
    second_sum = sum(second.values())

    max_id = max([int(i) for i in first.keys()] + [int(i) for i in second.keys()])
    for i in range(1, max_id +1):
        first_ret += [first.get(str(i), 0) / first_sum]
        second_ret += [second.get(str(i), 0) / second_sum]

    return first_ret, second_ret


def __combine_terms(docs):
    ret = {}
    for doc in docs:
        for term, count in doc.items():
            if term not in ret:
                ret[term] = 0
            ret[term] += count
    return ret


def build_probability_arrays_terms(first, second):
    first_ret = []
    second_ret = []
    first_sum = sum(first.values())
    second_sum = sum(second.values())

    terms = set([i for i in first.keys()] + [i for i in second.keys()])

    for i in sorted(terms):
        first_ret += [first.get(str(i), 0) / first_sum]
        second_ret += [second.get(str(i), 0) / second_sum]

    return first_ret, second_ret


def run_homogenity_experiments(json_line, num_splits=10):
    batches = combine_retrieved_documents_to_batches(json_line, num_splits)

    for batch in batches:
        if sum(batch.values()) == 0:
            return []

    topic = json.loads(json_line)['topic']
    ret = []

    for slice_ids in ALL_PERMUTATIONS:
        parts = [[], []]

        for i in range(num_splits):
            parts[slice_ids[i]] += [batches[i]]

        assert len(parts) == 2
        assert len(parts[0]) == 5
        assert len(parts[1]) == 5

        part_0 = __combine_terms(parts[0])
        part_1 = __combine_terms(parts[1])

        part_0, part_1 = build_probability_arrays_terms(part_0, part_1)

        ret += [{
            'topic': topic,
            'jensen_shannon_distance': jensen_shannon_distance(part_0, part_1),
        }]

    return ret


def evaluate_retrieval_homogenity(directory, index, retrieval_model, processes=20):
    from multiprocessing import Pool
    with Pool(processes=processes) as pool:
        lines = [i for i in open(directory + '/run.' + index + '.' + retrieval_model + '.txt.jsonl')]
        tmp_ret = [i for i in pool.map(run_homogenity_experiments, lines, 1)]
        ret = []
        for li in tmp_ret:
            for i in li:
                i['index'] = index
                i['retrieval_model'] = retrieval_model
                ret += [i]

        return ret


def combine_retrieved_documents_to_batches(json_line, num_splits=10):
    json_line = json.loads(json_line)
    documents = [i['documentRepresentation'] for i in json_line['documents'] if i is not None and 'documentRepresentation' in i and
 i['documentRepresentation'] is not None and len(i['documentRepresentation'].keys()) >= 2]
    random.Random(num_splits).shuffle(documents)
    
    return [__combine_terms(i) for i in np.array_split(documents, num_splits)]


def jensen_shannon_distance(p, q):
    from scipy.spatial.distance import jensenshannon
    return jensenshannon(p,q)

def kullback_leibler_divergence(p,q):
    from scipy.stats import entropy
    assert p != None
    assert q != None # scipy.stats.entropy calculate KL-divergence if p and q are both not none

    return entropy(p, q)

