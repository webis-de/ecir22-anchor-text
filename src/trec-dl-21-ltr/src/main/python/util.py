def write_run_file(model_file, feature_file, system_name, out_file):
    with open(out_file, 'w') as f:
        f.write(run_file_content(model_file, feature_file, system_name))

def run_file_content(model_file, feature_file, system_name):
    from lightgbm import Booster
    model = Booster(model_file=model_file)
    ret = ''

    to_rerank = load_ranking_data(feature_file)
    for topic_to_rerank in to_rerank:
        doc_to_score = []
        y_predictions = model.predict(topic_to_rerank['documentFeatures'])
        assert len(y_predictions) == len(topic_to_rerank['documents'])

        for i in range(len(topic_to_rerank['documents'])):
            doc_to_score += [{'documentId': topic_to_rerank['documents'][i], 'score': y_predictions[i]}]

        doc_to_score = sorted(doc_to_score, key=lambda i: i['score'], reverse=True)
        ret += topic_as_run_file(topic_to_rerank['topic'], doc_to_score, system_name)

    return ret.strip()

def topic_as_run_file(topic_nr, doc_to_score, system_name):
    ret = ''

    for i in range(len(doc_to_score)):
        ret += topic_nr + ' Q0 ' + doc_to_score[i]['documentId'] + ' ' + str(i+1) + ' ' + str(doc_to_score[i]['score']) + ' ' + system_name + '\n'

    return ret

def load_ranking_data(file_name):
    import json
    ret = []
    with open(file_name, 'r') as f:
        for l in f:
            l = json.loads(l)

            ret += [{
                'topic': l['topic'],
                'documents': [i['documentId'] for i in l['documents']],
                'documentFeatures': as_np_array(l['documents']),
            }]
    return ret


def as_np_array(docs):
    import numpy as np
    ret = []
    for doc in docs:
        last_feature=1
        features = doc['features']
        # relevance label and query id
        ret_features = [-100, -100]
        for feature_num in features:
            if (int(feature_num) -1) != last_feature:
                raise ValueError('Could not handle...')

            ret_features += [features[feature_num]]
            last_feature = int(feature_num)
        ret += [ret_features]

    return np.array(ret)

