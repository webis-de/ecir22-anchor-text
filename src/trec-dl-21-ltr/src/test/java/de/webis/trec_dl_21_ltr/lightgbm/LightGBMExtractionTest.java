package de.webis.trec_dl_21_ltr.lightgbm;

import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Arrays;
import java.util.HashSet;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;

import org.approvaltests.Approvals;
import org.junit.Assert;
import org.junit.Test;
import org.mockito.Matchers;
import org.mockito.Mockito;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;

import com.fasterxml.jackson.databind.ObjectMapper;

import io.anserini.eval.RelevanceJudgments;
import lombok.SneakyThrows;

public class LightGBMExtractionTest {

  private static final String DOCUMENT_FEATURES_FILE = "src/test/resources/feature-extraction/ms-marco-document-features.jsonl";
  private static final String QUERY_FEATURES_FILE = "src/test/resources/feature-extraction/ms-marco-query-features.jsonl";
  private static final String QUERY_DOCUMENT_FEATURES_DIR = "src/test/resources/feature-extraction/";

  @Test
  public void assert50DistinctFeaturesAreUsed() {
    Assert.assertEquals(50, new HashSet<>(App.FEATURES).size());
  }
  
  @Test
  public void assert121DistinctV1FeaturesAreUsed() {
    Assert.assertEquals(121, new HashSet<>(App.FEATURES_V1).size());
  }
  
  @Test
  public void assert112DistinctV1WithoutOrcasFeaturesAreUsed() {
    Assert.assertEquals(112, new HashSet<>(App.FEATURES_V1_NO_ORCAS).size());
  }
 
  @Test
  public void assert49DistinctV1WithoutAnchorFeaturesAreUsed() {
    Assert.assertEquals(49, new HashSet<>(App.FEATURES_V1_NO_ANCHOR).size());
  }
  
  @Test
  public void assert40DistinctV1WithoutAnchorWithoutOrcasFeaturesAreUsed() {
    Assert.assertEquals(40, new HashSet<>(App.FEATURES_V1_NO_ANCHOR_NO_ORCAS).size());
  }
  
  @Test
  @SneakyThrows
  public void approveFeatures() {
    Map<Integer, String> actual = new LinkedHashMap<>();
    for(int i=0; i< App.FEATURES.size(); i++) {
      actual.put(i+2, App.FEATURES.get(i));
    }

    Approvals.verify(new ObjectMapper().writeValueAsString(actual));
  }

  @Test
  @SneakyThrows
  public void approveFeaturesWithoutAnchors() {
    Map<Integer, String> actual = new LinkedHashMap<>();
    for(int i=0; i< App.FEATURES_NO_ANCHOR.size(); i++) {
      actual.put(i+2, App.FEATURES_NO_ANCHOR.get(i));
    }

    Approvals.verify(new ObjectMapper().writeValueAsString(actual));
  }

  @Test
  @SneakyThrows
  public void approveFeaturesV1() {
    Map<Integer, String> actual = new LinkedHashMap<>();
    for(int i=0; i< App.FEATURES_V1.size(); i++) {
      actual.put(i+2, App.FEATURES_V1.get(i));
    }

    Approvals.verify(new ObjectMapper().writeValueAsString(actual));
  }

  @Test
  @SneakyThrows
  public void approveFeaturesV1NoOrcas() {
    Map<Integer, String> actual = new LinkedHashMap<>();
    for(int i=0; i< App.FEATURES_V1_NO_ORCAS.size(); i++) {
      actual.put(i+2, App.FEATURES_V1_NO_ORCAS.get(i));
    }

    Approvals.verify(new ObjectMapper().writeValueAsString(actual));
  }

  @Test
  @SneakyThrows
  public void approveFeaturesV1NoAnchorNoOrcas() {
    Map<Integer, String> actual = new LinkedHashMap<>();
    for(int i=0; i< App.FEATURES_V1_NO_ANCHOR_NO_ORCAS.size(); i++) {
      actual.put(i+2, App.FEATURES_V1_NO_ANCHOR_NO_ORCAS.get(i));
    }

    Approvals.verify(new ObjectMapper().writeValueAsString(actual));
  }

  @Test
  @SneakyThrows
  public void approveFeaturesV1NoAnchor() {
    Map<Integer, String> actual = new LinkedHashMap<>();
    for(int i=0; i< App.FEATURES_V1_NO_ANCHOR.size(); i++) {
      actual.put(i+2, App.FEATURES_V1_NO_ANCHOR.get(i));
    }

    Approvals.verify(new ObjectMapper().writeValueAsString(actual));
  }  

  @Test
  public void testRunFile() {
    Integer relevanceLabel = null;
    Map<String, Object> features = Map.of(
      "f1", 0.01,
      "f2", 0.02,
      "f3", 0.03
    );
    
    String expected = "0 1:-4 2:0.02 3:0.01";
    String queryId = "-4";
    String documentId = "doc-2";
    
    String actual = LightGBMExtraction.formatLine(queryId, documentId, features, Arrays.asList("f2","f1"), relevanceLabel);
    
    Assert.assertEquals(expected, actual);
  }
  
  @Test
  public void testRunFile2() {
    Integer relevanceLabel = null;
    Map<String, Object> features = Map.of(
      "f1", 0.01,
      "f2", 0.02,
      "f3", 0.03
    );
    
    String expected = "0 1:queryId 2:0.01 3:0.02 4:0.03";
    String queryId = "queryId";
    String documentId = "doc";
    
    String actual = LightGBMExtraction.formatLine(queryId, documentId, features, Arrays.asList("f1","f2", "f3"), relevanceLabel);
    
    Assert.assertEquals(expected, actual);
  }
  
  @Test(expected = RuntimeException.class)
  public void testExtractionOfNonExistingFeature() {
    Integer relevanceLabel = null;
    Map<String, Object> features = Map.of(
      "f1", 0.01,
      "f2", 0.02,
      "f3", 0.03
    );
    String queryId = "queryId";
    String documentId = "doc";
    
    LightGBMExtraction.formatLine(queryId, documentId, features, Arrays.asList("f1","f2", "f4"), relevanceLabel);
  }

  @Test
  @SneakyThrows
  public void approveTestFeatureFile() {
    String outDir = Files.createTempDirectory("lightgbm-test").toAbsolutePath().toString();
    LightGBMConfiguration conf = new LightGBMConfiguration();
    LightGBMExtraction extract = LightGBMExtraction.builder()
    	.documentFeaturesFile(DOCUMENT_FEATURES_FILE)
    	.queryFeaturesFile(QUERY_FEATURES_FILE)
    	.queryDocumentFeaturesDirectory(QUERY_DOCUMENT_FEATURES_DIR)
    	.features(App.FEATURES)
    	.lightgbmConfig(conf)
    	.relevanceJudgments(relevanceJudgments())
    	.trainingTopics(dummyTopics())
    	.validationTopics(Arrays.asList(dummyTopics()))
    	.testTopics(dummyTopics())
        .build();
    extract.buildLightGbmDir(outDir);
    
    Approvals.verify(Files.readString(Paths.get(outDir).resolve("2021_trec_dl_test")));
  }

  @Test
  @SneakyThrows
  public void approveTestRerankFile() {
    String outDir = Files.createTempDirectory("lightgbm-test").toAbsolutePath().toString();
    LightGBMConfiguration conf = new LightGBMConfiguration();
    LightGBMExtraction extract = LightGBMExtraction.builder()
    	.documentFeaturesFile(DOCUMENT_FEATURES_FILE)
    	.queryFeaturesFile(QUERY_FEATURES_FILE)
    	.queryDocumentFeaturesDirectory(QUERY_DOCUMENT_FEATURES_DIR)
    	.features(App.FEATURES)
    	.lightgbmConfig(conf)
    	.relevanceJudgments(relevanceJudgments())
    	.trainingTopics(dummyTopics())
    	.validationTopics(Arrays.asList(dummyTopics()))
    	.testTopics(dummyTopics())
        .build();
    extract.buildLightGbmDir(outDir);
    
    Approvals.verify(Files.readString(Paths.get(outDir).resolve("2021_trec_dl_test.rerank")));
  }
  
  @Test
  @SneakyThrows
  public void approveValidationFeatureFile() {
    String outDir = Files.createTempDirectory("lightgbm-test").toAbsolutePath().toString();
    LightGBMConfiguration conf = new LightGBMConfiguration();
    LightGBMExtraction extract = LightGBMExtraction.builder()
    	.documentFeaturesFile(DOCUMENT_FEATURES_FILE)
    	.queryFeaturesFile(QUERY_FEATURES_FILE)
    	.queryDocumentFeaturesDirectory(QUERY_DOCUMENT_FEATURES_DIR)
    	.features(App.FEATURES)
    	.lightgbmConfig(conf)
    	.relevanceJudgments(relevanceJudgments())
    	.trainingTopics(dummyTopics())
    	.validationTopics(Arrays.asList(dummyTopics()))
    	.testTopics(dummyTopics())
        .build();
    extract.buildLightGbmDir(outDir);
    
    Approvals.verify(Files.readString(Paths.get(outDir).resolve("docv2_dev1")));
  }

  @Test
  @SneakyThrows
  public void approveTrainFeatureFile() {
    String outDir = Files.createTempDirectory("lightgbm-test").toAbsolutePath().toString();
    LightGBMConfiguration conf = new LightGBMConfiguration();
    LightGBMExtraction extract = LightGBMExtraction.builder()
    	.documentFeaturesFile(DOCUMENT_FEATURES_FILE)
    	.queryFeaturesFile(QUERY_FEATURES_FILE)
    	.queryDocumentFeaturesDirectory(QUERY_DOCUMENT_FEATURES_DIR)
    	.features(App.FEATURES)
    	.lightgbmConfig(conf)
    	.relevanceJudgments(relevanceJudgments())
    	.trainingTopics(dummyTopics())
    	.validationTopics(Arrays.asList(dummyTopics()))
    	.testTopics(dummyTopics())
        .build();
    extract.buildLightGbmDir(outDir);
    
    Approvals.verify(Files.readString(Paths.get(outDir).resolve("docv2_train")));
  }
  
  @Test
  @SneakyThrows
  public void approveTrainFeatureFileMsMarcoV1() {
    String outDir = Files.createTempDirectory("lightgbm-test").toAbsolutePath().toString();
    LightGBMConfiguration conf = new LightGBMConfiguration();
    LightGBMExtraction extract = LightGBMExtraction.builder()
    	.documentFeaturesFile("src/test/resources/feature-extraction/ms-marco-v1-document-features.jsonl")
    	.queryFeaturesFile("src/test/resources/feature-extraction/ms-marco-v1-query-features.jsonl")
    	.queryDocumentFeaturesDirectory(QUERY_DOCUMENT_FEATURES_DIR)
    	.features(App.FEATURES_V1)
    	.lightgbmConfig(conf)
    	.relevanceJudgments(relevanceJudgments())
    	.trainingTopics(dummyTopicsV1())
    	.validationTopics(Arrays.asList(dummyTopicsV1()))
    	.testTopics(dummyTopicsV1())
        .build();
    extract.buildLightGbmDir(outDir);
    
    Approvals.verify(Files.readString(Paths.get(outDir).resolve("docv2_train")));
  }
  
  private Map<Integer, Set<String>> dummyTopics() {
    return Map.of(
      1000006, Set.of("msmarco_doc_09_1282503587", "msmarco_doc_14_800591714"),
      1000412, Set.of("msmarco_doc_21_243726663", "msmarco_doc_21_243753677")
    );
  }
  
  private Map<Integer, Set<String>> dummyTopicsV1() {
    return Map.of(
      1090311, Set.of("D99747", "D99748")
    );
  }

  private RelevanceJudgments relevanceJudgments() {
    RelevanceJudgments relevanceJudgments = Mockito.mock(RelevanceJudgments.class);
    Mockito.when(relevanceJudgments.getRelevanceGrade(Matchers.any(), Matchers.any())).thenAnswer(new Answer<Integer>() {
      @Override
      public Integer answer(InvocationOnMock invocation) throws Throwable {
        return Arrays.asList("msmarco_doc_09_1282503587", "msmarco_doc_21_243753677", "D99747").contains(invocation.getArguments()[1]) ? 1 : 0;
      }
    });
	Mockito.when(relevanceJudgments.isDocJudged(Matchers.any(), Matchers.any())).thenReturn(true);
    
    return relevanceJudgments;
  }
}
