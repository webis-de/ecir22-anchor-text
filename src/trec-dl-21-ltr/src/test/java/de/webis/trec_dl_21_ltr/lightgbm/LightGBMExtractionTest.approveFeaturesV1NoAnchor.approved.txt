{"2":"body-bm25-similarity","3":"body-f2exp-similarity","4":"body-spl-similarity","5":"body-pl2-similarity","6":"body-ql-similarity","7":"body-qljm-similarity","8":"body-f2log-similarity","9":"body-tf-idf-similarity","10":"body-tf-similarity","11":"url-bm25-similarity","12":"url-f2exp-similarity","13":"url-spl-similarity","14":"url-pl2-similarity","15":"url-ql-similarity","16":"url-qljm-similarity","17":"url-f2log-similarity","18":"url-tf-idf-similarity","19":"url-tf-similarity","20":"title-bm25-similarity","21":"title-f2exp-similarity","22":"title-spl-similarity","23":"title-pl2-similarity","24":"title-ql-similarity","25":"title-qljm-similarity","26":"title-f2log-similarity","27":"title-tf-idf-similarity","28":"title-tf-similarity","29":"orcas-bm25-similarity","30":"orcas-f2exp-similarity","31":"orcas-spl-similarity","32":"orcas-pl2-similarity","33":"orcas-ql-similarity","34":"orcas-qljm-similarity","35":"orcas-f2log-similarity","36":"orcas-tf-idf-similarity","37":"orcas-tf-similarity","38":"feature_is_5w1h_query","39":"feature_query_length_in_tokens","40":"feature_gpe_entities","41":"feature_org_entities","42":"feature_person_entities","43":"feature_url_length","44":"feature_slashes_in_url","45":"feature_dots_in_host","46":"feature_body_length","47":"feature_title_length","48":"feature_pagerank","49":"feature_harmonic_mean","50":"feature_alexa_rank"}